enterprise/api/v1/views/default_enterprise_enrollments.py

```diff
@@ -0,0 +1,270 @@
+
+        Arguments:
+            enterprise_enrollments: List of tuples containing course_id and whether enrollment is audit.
+
+        Returns:
+            Tuple of two lists containing course ids: (enrolled_non_audit, enrolled_audit)
+        """
+        enrolled_non_audit = []
+        enrolled_audit = []
+        for enrollment in enterprise_course_enrollments:
+            (enrolled_audit if enrollment.is_audit_enrollment else enrolled_non_audit).append(enrollment)
+        return enrolled_non_audit, enrolled_audit
+
+    def _get_audit_modes(self):
+        """
+        Get the configured audit modes from settings.
+        """
+        return getattr(settings, 'ENTERPRISE_COURSE_ENROLLMENT_AUDIT_MODES', ['audit', 'honor'])
+
+    def _partition_default_enrollment_intentions_by_enrollment_status(
+        self,
+        default_enrollment_intentions,
+        enrolled_non_audit,
+        enrolled_audit,
+    ):
+        """
+        Partition default enrollment intentions into enrollable and non-enrollable.
+
+        Arguments:
+            default_enrollment_intentiosn: List of default enrollment intentions.
+            enrolled_non_audit: List of course runs in which the learner is enrolled in non-audit mode.
+            enrolled_audit: List of course runs in which the learner is enrolled in audit mode.
+
+        Returns:
+            Tuple of three lists: (already_enrolled, needs_enrollment_enrollable, needs_enrollment_not_enrollable)
+        """
+        already_enrolled = []
+        needs_enrollment_enrollable = []
+        needs_enrollment_not_enrollable = []
+
+        for intention in default_enrollment_intentions:
+            has_applicable_catalogs = intention.applicable_enterprise_catalog_uuids
+            non_audit_enrollment = next(
+                (enrollment for enrollment in enrolled_non_audit if enrollment.course_id == intention.course_run_key),
+                None
+            )
+            audit_enrollment = next(
+                (enrollment for enrollment in enrolled_audit if enrollment.course_id == intention.course_run_key),
+                None
+            )
+
+            if non_audit_enrollment and non_audit_enrollment.is_active:
+                # Learner is already enrolled  (is_active=True) in non-audit mode for this course run
+                already_enrolled.append((intention, non_audit_enrollment))
+                continue
+
+            if not intention.is_course_run_enrollable:
+                # Course run is not enrollable
+                needs_enrollment_not_enrollable.append((intention, audit_enrollment))
+                continue
+
+            has_non_audit_mode_for_course_run = intention.best_mode_for_course_run not in self._get_audit_modes()
+            is_audit_enrollment_with_non_audit_modes = audit_enrollment and has_non_audit_mode_for_course_run
+
+            if is_audit_enrollment_with_non_audit_modes and has_applicable_catalogs:
```

Comment:
Nice job commenting this and making it easier to reason about. Maybe it's worth adding a note at the top here that the order of evaluation matters and `elif` conditions shouldn't be moved around unless you really know what you're doing.